pretyping/nativenorm.ml

```diff
@@ -130,16 +130,19 @@ let construct_of_constr_const env sigma tag typ =
 
 let construct_of_constr_block = construct_of_constr false
 
-let build_branches_type env sigma mib mip (ind,u) params p =
+let get_case_annot decls =
+  Array.map_of_list (fun decl -> get_annot decl) (List.rev decls)
+
+let build_branches_type env sigma mib mip (ind,u) params (pctx, p) =
   let rtbl = mip.mind_reloc_tbl in
   let paramsl = Array.to_list params in
   (* [build_one_branch i cty] construit le type de la ieme branche (commence
      a 0) et les lambda correspondant aux realargs *)
-  let build_one_branch i =
+  let p = it_mkLambda_or_LetIn p pctx in (* TODO: prevent useless cut? *)
```

Comment:
Have you got a commit preventing it? I'm interested in seeing if it makes any difference to the bench.

Reply:
I don't have the commit ready, but it's trivial. I kind of fear that it might be slower in the current implementation, though, because the trade-off is between reconstructing the types of the abstraction vs. substituting in the body. The latter is arbitrarily big, so it can be worse than type reconstruction. The trick is that *if the type is needed* by reification, the current implementation will perform the substitution anyways, so we're in the worst situation. Yet, not all types are needed for reification. I think that the best situation would require computing the reification type lazily *and* substituting (lazily) rather than reconstructing the abstracted types, but this is a little bit more invasive.